### ISSUE TITLE
Hashtbl.merge fails

### ISSUE THREAD

**Issue by alfa07** [janestreet/base#12]
This one fails:
```ocaml
open Base

let pr fmt = Caml.Printf.printf (Caml.(^^) fmt "\n")

let () =
  let h0 = Hashtbl.create (module String) () in
  (* This throws exception in merge *)
  let h1 = Hashtbl.create (module String) () in
  (* This one works: *)
  (* let h1 = Hashtbl.copy h0 in *)
  Hashtbl.add h0 "a" 1 |> ignore;
  Hashtbl.add h1 "a" 1 |> ignore;
  Hashtbl.add h1 "b" 2 |> ignore;
  let h2 = Hashtbl.merge h0 h1 ~f:(
    fun ~key values ->
      match values with
      | `Both (left, right) -> Some right
      | `Left left -> Some left
      | `Right right -> Some right
  )
  in
  pr "Testing base: %d" (Hashtbl.length h2)
```
I understand that a `module String` object is each time unique. How I can share it (apart from using `copy`)? I've tried to create it with `let` but I couldn't make it compile (what module type should I use?) What if I need to serialize, deserialize, and merge back?

**Comment by bmillwood** [comment]
Hmm. This works:

    let string_key = (module String : Hashtbl_intf.Key with type t = string) in
    let tbl = Hashtbl.create string_key

but it's a bit wordy. In `Core_kernel` and `Core` we have idioms that make this more convenient, but it looks like they're not available in `Base`. I'll let @diml comment on whether the above is the expected way of doing it.

**Comment by diml** [comment]
Basically we were starting to have an explosion of `Make_thingable` in Core and we wanted a lighter way of combining things in Base. I guess this is in one of the current drawback. Ideally, the compiler would just statically allocate `(module String)` once and things would work as expected, but this is not the case at the moment.

The workaround is to define the key type once:

```ocaml
let key = (module String : Hashtbl_intf.Key_plain with type t = string) in
...
```

**Comment by diml** [comment]
BTW, we don't have this problem for sets and maps since we have these special `comparators` where if we know that the types match, then we know that the comparison functions are the same. Maybe we should have something similar for hashable things

**Comment by diml** [comment]
I asked internally, in fact it seems that this check shouldn't exist: the new hash table can simply use one or the other of the two hash functions. It's not the same as for sets or maps where the ordering of elements matters

/cc @seliopou who worked on this module recently

**Comment by alfa07** [comment]
Awesome! Thank you for the replies! This thing I had trouble coming up myself:
```ocaml
let key = (module String : Hashtbl_intf.Key_plain with type t = string) in
```
Yeah, and check removal should solve the problem with whatever serialization one uses.

**Comment by seliopou** [comment]
This is fixed in master, and will be included in the next OPAM release.
